contracts/WooCrossChainRouter.sol

```diff
@@ -1,372 +1,446 @@
-
-    function quoteLayerZeroFee(
-        uint16 dstChainId,
-        address toToken,
-        uint256 refId,
-        uint256 dstMinToAmount,
-        address to
-    ) external view returns (uint256, uint256) {
-        bytes memory toAddress = abi.encodePacked(to);
-        bytes memory payloadData = abi.encode(
-            toToken, // to token
-            refId, // reference id
-            dstMinToAmount, // minToAmount on destination chain
-            to // to address
+        emit WooCrossSwapOnSrcChain(
+            refId,
+            msgSender,
+            to,
+            srcInfos.fromToken,
+            srcInfos.fromAmount,
+            srcInfos.minBridgeAmount,
+            bridgeAmount
         );
-        uint256 dstGas = (toToken == wooppQuoteTokens[dstChainId]) ? dstGasForNoSwapCall : dstGasForSwapCall;
-        return
-            stargateRouter.quoteLayerZeroFee(
-                dstChainId,
-                1, // https://stargateprotocol.gitbook.io/stargate/developers/function-types
-                toAddress,
-                payloadData,
-                IStargateRouter.lzTxObj(dstGas, 0, "0x")
-            );
     }
 
     function sgReceive(
-        uint16, /*_chainId*/
-        bytes memory, /*_srcAddress*/
-        uint256, /*_nonce*/
-        address _token,
+        uint16, // srcChainId
+        bytes memory, // srcAddress
+        uint256, // nonce
+        address bridgedToken,
         uint256 amountLD,
         bytes memory payload
     ) external override {
-        require(msg.sender == address(stargateRouter), "WooCrossChainRouter: INVALID_CALLER");
+        require(msg.sender == address(stargateRouter), "WooCrossChainRouterV2: INVALID_CALLER");
 
-        (address toToken, uint256 refId, uint256 minToAmount, address to) = abi.decode(
-            payload,
-            (address, uint256, uint256, address)
-        );
+        (uint256 refId, address to, address toToken, uint256 minToAmount) = _decodePayload(payload);
 
-        if (wooRouter.wooPool().quoteToken() != _token) {
-            // NOTE: The bridged token is not WooPP's quote token.
-            // So Cannot do the swap; just return it to users.
-            TransferHelper.safeTransfer(_token, to, amountLD);
+        // When bridged by ETH, the bridgedToken will be SGETH ERC20 address and send native token to address(this)
+        if (toToken == ETH_PLACEHOLDER_ADDR && bridgedToken == sgETHs[sglChainId]) {
+            TransferHelper.safeTransferETH(to, amountLD);
             emit WooCrossSwapOnDstChain(
                 refId,
                 msg.sender,
                 to,
-                _token,
+                ETH_PLACEHOLDER_ADDR,
                 amountLD,
                 toToken,
-                _token,
+                ETH_PLACEHOLDER_ADDR,
                 minToAmount,
                 amountLD
             );
             return;
         }
 
-        uint256 quoteAmount = amountLD;
+        if (toToken == bridgedToken) {
```

Comment:
is that possible here, the bridgedToken could be ETH_PLACEHOLDER_ADDR ?

Reply:
In `sgReceive`, bridgedToken wont be `ETH_PLACEHOLDER_ADDR`. Only ERC20 token address.

Reply:
If bridged by ETH on source chain, then sgRecevie`s bridgedToken will be [SGETH](https://arbiscan.io/address/0x82CbeCF39bEe528B5476FE6d1550af59a9dB6Fc0) address.